WordPress/Classes/ViewRelated/Media/AnimatedImageCache.swift

```diff
@@ -101,13 +85,8 @@ class AnimatedImageCache {
 
             let staticImage = UIImage(data: data)
             if let key = urlRequest.url {
-                let dataByteCost = AnimatedImageCache.byteCost(for: data)
-                AnimatedImageCache.cache.setObject(data as NSData, forKey: key as NSURL, cost: dataByteCost)
-
-                // Creating a static image from GIF data is an expensive op, so let's try to do it once...
-                let imageByteCost = AnimatedImageCache.byteCost(for: staticImage)
-                let imageKey = key.absoluteString + Constants.keyStaticImageSuffix
-                AnimatedImageCache.cache.setObject(staticImage as AnyObject, forKey: imageKey as AnyObject, cost: imageByteCost)
+                self.cacheData(data: data, url: key)
+                self.cacheStaticImage(url: key, image: staticImage)
```

Comment:
I couldn't figure out why both the data and the image are saved. The old (now deleted code) says that creating the static image is expensive, so it's cached alongside the data. But isn't the static image just one frame of the gif? If so, I'm not sure why it's creating the image with the entire gif's data. The only explanation I see is that `UIImage` only takes the first frame when the data is passed to it and discards the rest. Is this the case?
I will do test this today but wanted to ask here in case there's something obvious I'm missing.

Reply:
> "Creating a static image from GIF data is an expensive op" 

I don't think this comment is accurate. `UIImage(data:)` is a quick operation because it doesn't perform any decompression by default. All the expensive processing happens on demand when you try to display the image. The image-loading frameworks force the (expensive) decompression to happen in the background, which is one of the big reasons to use them. Slow decompression (sometimes called bitmapping) means lost frames on the main thread.

> But isn't the static image just one frame of the gif? If so, I'm not sure why it's creating the image with the entire gif's data. The only explanation I see is that UIImage only takes the first frame when the data is passed to it and discards the rest. Is this the case?

When you call `UIImage(data:)` with GIF data, it only decodes and displays the first frame of the image. It has to hold on to the original data for at least some time before the image is decoded. I don't know what sure for how long it holds on to the data source because the image-loading frameworks avoid this situation entirely.

Based on what I see in the code, this first frame is used as a preview while the framework (Gifu) prepares the GIF for animation in the background, which is a common strategy.

Reply:
I was pretty confused when I saw that some images were not animated. Here is the code snippet that is responsible for that:

```swift
if didVerifyDataSize {
    self.animate(data: animatedImageData, success: success)
} else {
    self.animatedGifData = nil
    success?()
}
```

The limit seems to be 20 MB in the post previews, which is pretty arbitrary. There is nothing that prevents the app from rendering larger images, except for the unreasonably large custom buffer sizes that I fixed [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/21156).

Reply:
Thanks for the info. It would be good to read more about this, so if you have any links for further reading, please send them along. 

> When you call `UIImage(data:)` with GIF data, it only decodes and displays the first frame of the image.

Makes sense, I read up a bit on gif compression and it says [here](https://en.wikipedia.org/wiki/GIF#:~:text=Within%20each%20frame%20the%20LZW%2Dcoded%20image%20data%20is%20arranged) that each gif frame is compressed individually.

> the image-loading frameworks avoid this situation entirely by pre-rendering the first frame in the background and creating the image _without_ a data source.

By this, do you mean that instead of feeding the `UIImage(data:)` the entire gif's data, the image-loading frameworks extract only pass the first frame's data to the `UIImage(data:)`?

> What's uncommon is to store these two separately.

It does seem odd to cache both the data and the image, since that would double the cache's memory size, right? If this is the case, I'm happy to leave this to another discussion since this PR changes the cache, not what's stored in the cache.

Reply:
>  extract only pass the first frame's data to the UIImage(data:)?

You can extract the first frame using [`CGImageSourceCreateImageAtIndex`](https://developer.apple.com/documentation/imageio/1465011-cgimagesourcecreateimageatindex) and create a `UIImage` with a `CGImage`.

> since that would double the cache's memory size, right? 

Not necessarily. They should ultimately point to the same buffer in memory, so it should be OK. This PR does not change the existing behavior anyway.

The app should not even worry about these details; a framework should handle it.

Reply:
> They should ultimately point to the same buffer in memory, so it should be OK.

Ok

> This PR does not change the existing behavior anyway.

I agree 👍